File: integration/feature/test-quick/resources/app/src/MyString.scala
package app

import lib.*

final case class MyString(val value: String)

object MyString {
  
  given gCombinator: Combinator[MyString] = new Combinator[MyString] {
    def combine(a: MyString, b: MyString): MyString = MyString(a.value + b.value)
  }

  given gDefaultValue: DefaultValue[MyString] = new DefaultValue[MyString] {
    def defaultValue: MyString = MyString("")
  }

  def combine(a: MyString, b: MyString, c: MyString): MyString = {
    val temp = gCombinator.combine(a, b)
    gCombinator.combine(temp, c)
  }

  def defaultValue: MyString = gDefaultValue.defaultValue

}

File: integration/feature/test-quick-java/resources/app/src/MyString.java
package app;

import lib.Combinator;
import lib.DefaultValue;

public class MyString {
    private final String value;

    public MyString(String value) {
        this.value = value;
    }

    public String getValue() {
        return value;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof MyString) {
            return value.equals(((MyString) obj).value);
        }
        return false;
    }

    @Override
    public int hashCode() {
        return value.hashCode();
    }

    @Override
    public String toString() {
        return "MyString(" + value + ")";
    }

    public static final Combinator<MyString> COMBINATOR = new Combinator<MyString>() {
        @Override
        public MyString combine(MyString a, MyString b) {
            return new MyString(a.getValue() + b.getValue());
        }
    };

    public static final DefaultValue<MyString> DEFAULT_VALUE = new DefaultValue<MyString>() {
        @Override
        public MyString defaultValue() {
            return new MyString("");
        }
    };

    public static MyString combine(MyString a, MyString b, MyString c) {
        return COMBINATOR.combine2(a, b, c);
    }

    public static MyString getDefaultValue() {
        return DEFAULT_VALUE.defaultValue();
    }
}
